airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_incremental.py

```diff
@@ -249,25 +279,139 @@ async def test_state_with_abnormally_large_values(
+        record count for each state is the total record count minus the sum of record counts of all
+        previous states in the list.
+
+        Args:
+            states (list): A non-empty list of state objects.
+
+        Yields:
+            tuple: A tuple containing a state object and its expected record count.
+
+        Raises:
+            Exception: If the 'states' list is empty.
+        """
+        if not states:
+            # Raise an exception if the 'states' list is empty
+            raise Exception("`states` expected to be a non empty list.")
+
+        # Function to extract the record count from a state object
+        get_record_count = operator.attrgetter("state.sourceStats.recordCount")
+
+        # Calculate the total record count from all states
+        total_count = sum(map(get_record_count, states))
+
+        # Create an iterator of states and their reverse cumulative expected record counts
+        expected_record_count_per_state = zip(
+            states, [total_count - sum(map(get_record_count, states[: idx + 1])) for idx in range(len(states))]
+        )
+
+        # Yield expected record counts using the helper function
+        yield from self._get_expected_record_count_per_state(expected_record_count_per_state, total_count)
+
+    def _get_expected_record_count_per_state(self, expected_record_count_per_state, total_count, current_record_count_and_state=None):
+        """
+        Helper generator function to yield the expected record count per state.
+
+        This function is a recursive generator that processes the states and their respective
+        record counts, yielding them one by one and ensures that the states are yielded only when
+        their expected record count changes from the total record count.
+
+        Why do we need this:
+            When two or more consecutive expected record counts have the same value despite their states being different,
+            it means no records were produced for all of their respective states except the last one.
+            Applying those states will result in the same set of records during stream reading, leading to a test failure (diff check). That is why such states are skipped.
+
+        Args:
+            expected_record_count_per_state (iterator): An iterator of tuples, where each tuple contains a state object and its expected record count.
+            total_count (int): The total record count from all states combined.
+            current_record_count_and_state (tuple, optional): A tuple containing the current state object and its record count. Defaults to None.
+
+        Yields:
+            tuple: A tuple containing a state object and its expected record count.
+        """
+        if current_record_count_and_state is None:
+            # Initialize the current record count and state if not provided
+            current_record_count_and_state = next(expected_record_count_per_state)
```

Comment:
nit: the order is reversed. let's rename to `current_state_and_record_count`?

Reply:
Done. I renamed all the methods/variables so they reflect what they do/contain